compiler/rustc_expand/src/mbe/macro_rules.rs

```diff
@@ -953,6 +954,23 @@ fn check_matcher_core(
         for token in &last.tokens {
             if let TokenTree::MetaVarDecl(_, name, Some(kind)) = *token {
                 for next_token in &suffix_first.tokens {
+                    if let NonterminalKind::Pat2015 { inferred: true } = kind {
+                        if let TokenTree::Token(token) = next_token {
+                            if let BinOp(token) = token.kind {
+                                if let token::BinOpToken::Or = token {
+                                    sess.buffer_lint_with_diagnostic(
+                                        &OR_PATTERNS_BACK_COMPAT,
+                                        next_token.span(),
+                                        ast::CRATE_NODE_ID,
+                                        "detects usage of old versions of or patterns",
+                                        BuiltinLintDiagnostics::OrPatternsBackCompat(
+                                            "the pat2021 matcher will match new patterns, which include the | character. \
+                                Please use pat2015 to avoid breakage.".to_string()),
+                                    );
+                                }
+                            }
+                        }
+                    }
                     match is_in_follow(next_token, kind) {
```

Comment:
We should modify the follow set for Pat2021 (do I understand correctly that, by the time we get to this code, we either have Pat2015 or Pat2021?) to exclude `|`, too.